lib/valueflow.cpp

```diff
@@ -1596,52 +1613,107 @@ static bool valueFlowForward(Token * const               startToken,
+        for (Token* tok = const_cast<Token*>(start); tok != scope->classEnd; tok = tok->next()) {
+            while (!flowBariers.empty() && flowBariers.back().tok == tok)
+                flowBariers.pop_back();
+            Token * varTok;
+            if ((tok->str() == "=") && (!tok->astParent())) {
+                // Assignment
+
+                // Lhs should be a variable
+                if (!tok->astOperand1() || !tok->astOperand1()->varId()) {
+                    continue;
+                }
+                const unsigned int varid = tok->astOperand1()->varId();
+                const Token * tokEndOfAssignment = findEndOfStatement(tok);
+                flowBariers.push_back(FlowBarierType(varid, tokEndOfAssignment));
+                const Variable *var = tok->astOperand1()->variable();
+                if (!var || (!var->isLocal() && !var->isArgument())) {
+                    continue;
+                }
 
-            const Token * const endOfVarScope = var->typeStartToken()->scope()->classEnd;
+                const Token * const endOfVarScope = var->typeStartToken()->scope()->classEnd;
 
-            // Rhs values..
-            if (!tok->astOperand2() || tok->astOperand2()->values.empty())
-                continue;
+                // Rhs values..
+                if (!tok->astOperand2() || tok->astOperand2()->values.empty()) {
+                    continue;
+                }
 
-            std::list<ValueFlow::Value> values = tok->astOperand2()->values;
-            const bool constValue = tok->astOperand2()->isNumber();
+                std::list<ValueFlow::Value> values = tok->astOperand2()->values;
+                const bool constValue = tok->astOperand2()->isNumber();
 
-            // Static variable initialisation?
-            if (var->isStatic() && var->nameToken() == tok->astOperand1()) {
-                for (std::list<ValueFlow::Value>::iterator it = values.begin(); it != values.end(); ++it) {
-                    it->changeKnownToPossible();
+                // Static variable initialisation?
+                if (var->isStatic() && var->nameToken() == tok->astOperand1()) {
+                    for (std::list<ValueFlow::Value>::iterator it = values.begin(); it != values.end(); ++it) {
+                        it->changeKnownToPossible();
+                    }
                 }
-            }
 
-            // Skip RHS
-            const Token *nextExpression = tok->astOperand2();
-            for (;;) {
-                if (nextExpression->astOperand2())
-                    nextExpression = nextExpression->astOperand2();
-                else if (nextExpression->isUnaryPreOp())
-                    nextExpression = nextExpression->astOperand1();
-                else break;
-            }
-            nextExpression = nextExpression->next();
+                std::vector<FlowBarierType>::reverse_iterator foundFlowBarier = find_if(flowBariers.rbegin()+1, flowBariers.rend(), FlowBarierVaridPredicate(varid));
```

Comment:
I prefer full name std::find_if .. this is not portable there is not a "using namespace std;" anywhere.